src/engine.ml

```diff
@@ -1562,6 +1619,42 @@ let wrap_control state control_crypto needs_wkc key transport outs =
   in
   (session, transport, outs)
 
+let send_control_message s data =
+  let open Result.Syntax in
+  match s.control_crypto with
+  | `Static _ ->
+      (* XXX(reynir) :/ *)
+      Error `Not_ready
+  | #control_tls as control_crypto ->
+      let* channel_st, out =
+        match (s.state, s.channel.channel_st) with
+        | (Client Ready | Server Server_ready), Established (tls, keys) ->
+            let data' = Cstruct.create_unsafe (String.length data + 1) in
+            Cstruct.blit_from_string data 0 data' 0 (String.length data);
+            Cstruct.set_uint8 data' (String.length data) 0;
+            let* tls, out =
+              (* reynir: I *think* it only returns [None] when not established
+                 or after write is shutdown. *)
```

Comment:
yes :) that assumption is shared with me -- the send_application_data shouldn't return None if the channel is ready.